
# Machine Learning Based Recommendation Systems
##  Model-based Collaborative Filtering Systems
## SVD Matrix Factorization

In [2]:
import numpy as np
import pandas as pd

import sklearn
from sklearn.decomposition import TruncatedSVD

The MovieLens dataset was collected by the GroupLens Research Project at the University of Minnesota. You can download the dataset for this demostration at the following URL: https://grouplens.org/datasets/movielens/100k/

### Preparing the data

In [3]:
columns = ['user_id', 'item_id', 'rating', 'timestamp']
frame = pd.read_csv('ml-100k/u.data', sep='\t', names=columns)
frame.head()

user_id  item_id  rating  timestamp
0      196      242       3  881250949
1      186      302       3  891717742
2       22      377       1  878887116
3      244       51       2  880606923
4      166      346       1  886397596

In [4]:
frame.describe()

user_id        item_id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

Note count	100000.00000 from 1.7M movies is very inefficient

In [5]:
columns = ['item_id', 'movie title', 'release date', 'video release date', 'IMDb URL', 'unknown', 'Action', 'Adventure',
          'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror',
          'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']

movies = pd.read_csv('ml-100k/u.item', sep='|', names=columns, encoding='latin-1')
movie_names = movies[['item_id', 'movie title']]
movie_names.head()

item_id        movie title
0        1   Toy Story (1995)
1        2   GoldenEye (1995)
2        3  Four Rooms (1995)
3        4  Get Shorty (1995)
4        5     Copycat (1995)

In [49]:
combined_movies_data = pd.merge(frame, movie_names, on='item_id')
combined_movies_data.head()

user_id  item_id  rating  timestamp   movie title
0      196      242       3  881250949  Kolya (1996)
1       63      242       3  875747190  Kolya (1996)
2      226      242       5  883888671  Kolya (1996)
3      154      242       3  879138235  Kolya (1996)
4      306      242       5  876503793  Kolya (1996)

In [50]:
# Group by item_id sorting count of 'rating'
combined_movies_data.groupby('item_id')['rating'].count().sort_values(ascending=False).head()

item_id
50     583
258    509
100    508
181    507
294    485
Name: rating, dtype: int64

In [51]:
# Get movie name of item_id == 50
filter = combined_movies_data['item_id']==50
combined_movies_data[filter]['movie title'].unique()

array(['Star Wars (1977)'], dtype=object)

### Building a Utility Matrix

In [52]:
rating_crosstab = combined_movies_data.pivot_table(values='rating',index='user_id', columns='movie title').fillna(0)
rating_crosstab.head()

movie title  'Til There Was You (1997)  1-900 (1994)  101 Dalmatians (1996)  \
user_id                                                                       
1                                  0.0           0.0                    2.0   
2                                  0.0           0.0                    0.0   
3                                  0.0           0.0                    0.0   
4                                  0.0           0.0                    0.0   
5                                  0.0           0.0                    2.0   

movie title  12 Angry Men (1957)  187 (1997)  2 Days in the Valley (1996)  \
user_id                                                                     
1                            5.0         0.0                          0.0   
2                            0.0         0.0                          0.0   
3                            0.0         2.0                          0.0   
4                            0.0         0.0                          0.0   
5                            0.0         0.0                          0.0   

movie title  20,000 Leagues Under the Sea (1954)  \
user_id                                            
1                                            3.0   
2                                            0.0   
3                                            0.0   
4                                            0.0   
5                                            0.0   

movie title  2001: A Space Odyssey (1968)  \
user_id                                     
1                                     4.0   
2                                     0.0   
3                                     0.0   
4                                     0.0   
5                                     4.0   

movie title  3 Ninjas: High Noon At Mega Mountain (1998)  \
user_id                                                    
1                                                    0.0   
2                                                    1.0   
3                                                    0.0   
4                                                    0.0   
5                                                    0.0   

movie title  39 Steps, The (1935)                 ...                  \
user_id                                           ...                   
1                             0.0                 ...                   
2                             0.0                 ...                   
3                             0.0                 ...                   
4                             0.0                 ...                   
5                             0.0                 ...                   

movie title  Yankee Zulu (1994)  Year of the Horse (1997)  \
user_id                                                     
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
5                           0.0                       0.0   

movie title  You So Crazy (1994)  Young Frankenstein (1974)  \
user_id                                                       
1                            0.0                        5.0   
2                            0.0                        0.0   
3                            0.0                        0.0   
4                            0.0                        0.0   
5                            0.0                        4.0   

movie title  Young Guns (1988)  Young Guns II (1990)  \
user_id                                                
1                          3.0                   0.0   
2                          0.0                   0.0   
3                          0.0                   0.0   
4                          0.0                   0.0   
5                          0.0                   0.0   

movie title  Young Poisoner's

In [53]:
rating_crosstab = combined_movies_data.pivot_table(values='rating',index='user_id', columns='movie title', fill_value=0)
rating_crosstab.head()

movie title  'Til There Was You (1997)  1-900 (1994)  101 Dalmatians (1996)  \
user_id                                                                       
1                                    0             0                      2   
2                                    0             0                      0   
3                                    0             0                      0   
4                                    0             0                      0   
5                                    0             0                      2   

movie title  12 Angry Men (1957)  187 (1997)  2 Days in the Valley (1996)  \
user_id                                                                     
1                              5           0                            0   
2                              0           0                            0   
3                              0           2                            0   
4                              0           0                            0   
5                              0           0                            0   

movie title  20,000 Leagues Under the Sea (1954)  \
user_id                                            
1                                              3   
2                                              0   
3                                              0   
4                                              0   
5                                              0   

movie title  2001: A Space Odyssey (1968)  \
user_id                                     
1                                       4   
2                                       0   
3                                       0   
4                                       0   
5                                       4   

movie title  3 Ninjas: High Noon At Mega Mountain (1998)  \
user_id                                                    
1                                                      0   
2                                                      1   
3                                                      0   
4                                                      0   
5                                                      0   

movie title  39 Steps, The (1935)                 ...                  \
user_id                                           ...                   
1                               0                 ...                   
2                               0                 ...                   
3                               0                 ...                   
4                               0                 ...                   
5                               0                 ...                   

movie title  Yankee Zulu (1994)  Year of the Horse (1997)  \
user_id                                                     
1                             0                         0   
2                             0                         0   
3                             0                         0   
4                             0                         0   
5                             0                         0   

movie title  You So Crazy (1994)  Young Frankenstein (1974)  \
user_id                                                       
1                              0                          5   
2                              0                          0   
3                              0                          0   
4                              0                          0   
5                              0                          4   

movie title  Young Guns (1988)  Young Guns II (1990)  \
user_id                                                
1                            3                     0   
2                            0                     0   
3                            0                     0   
4                            0                     0   
5                            0                     0   

movie title  Young Poisoner's

### Transposing the Matrix

In [54]:
rating_crosstab.shape

(943, 1664)

In [55]:
X = rating_crosstab.T
X.shape

(1664, 943)

### Decomposing the Matrix: 
SVD see https://fenix.tecnico.ulisboa.pt/downloadFile/3779576344458/singular-value-decomposition-fast-track-tutorial.pdf  
SVD to reduce/compress information in large matrix(1664, 943) to smaller matrix(1664, 12) 

In [56]:
# n_components: hyper parameter, need to adjust at model evaluation
SVD = TruncatedSVD(n_components=12, random_state=17)
resultant_matrix = SVD.fit_transform(X)
resultant_matrix.shape

(1664, 12)

In [57]:
resultant_matrix[0]

array([ 1.03999361,  0.6598845 ,  0.04568944,  0.81444724,  0.16747427,
       -0.98316366, -0.3198289 ,  0.3067836 , -0.03084219, -0.75356126,
       -0.30152968, -0.53651145])

### Generating a Correlation Matrix

In [58]:
corr_mat = np.corrcoef(resultant_matrix)
corr_mat

array([[ 1.        , -0.10298113,  0.52210159, ...,  0.39854553,
         0.22143017,  0.5039286 ],
       [-0.10298113,  1.        ,  0.06549218, ...,  0.16134137,
         0.5091753 ,  0.23355053],
       [ 0.52210159,  0.06549218,  1.        , ...,  0.7658073 ,
         0.44348034,  0.19721751],
       ..., 
       [ 0.39854553,  0.16134137,  0.7658073 , ...,  1.        ,
         0.18088492,  0.10342131],
       [ 0.22143017,  0.5091753 ,  0.44348034, ...,  0.18088492,
         1.        ,  0.18524109],
       [ 0.5039286 ,  0.23355053,  0.19721751, ...,  0.10342131,
         0.18524109,  1.        ]])

In [59]:
corr_mat.shape

(1664, 1664)

In [60]:
corr_mat[0]

array([ 1.        , -0.10298113,  0.52210159, ...,  0.39854553,
        0.22143017,  0.5039286 ])

### Isolating Star Wars From the Correlation Matrix

In [67]:
movie_names = rating_crosstab.columns
movies_list = list(movie_names)
star_wars = movies_list.index('Star Wars (1977)')
star_wars

1398

In [68]:
corr_star_wars = corr_mat[1398]
corr_star_wars.shape

(1664,)

### Recommending a Highly Correlated Movie

In [69]:
list(movie_names[(corr_star_wars<1.0) & ( corr_star_wars > 0.9)])

['Die Hard (1988)',
 'Empire Strikes Back, The (1980)',
 'Fugitive, The (1993)',
 'Raiders of the Lost Ark (1981)',
 'Return of the Jedi (1983)',
 'Terminator 2: Judgment Day (1991)',
 'Terminator, The (1984)',
 'Toy Story (1995)']

Find something very similar starwars

In [71]:
list(movie_names[(corr_star_wars<1.0) & ( corr_star_wars > 0.95)])

['Return of the Jedi (1983)']

Something NOT similar

In [74]:
# something NOT similar
list(movie_names[(corr_star_wars<0.2) & ( corr_star_wars < 0.9)])

['Afterglow (1997)',
 'American Strays (1996)',
 'Angela (1995)',
 'Anna Karenina (1997)',
 'Assignment, The (1997)',
 'August (1996)',
 'Ayn Rand: A Sense of Life (1997)',
 'B. Monkey (1998)',
 'Babyfever (1994)',
 'Babysitter, The (1995)',
 'Baton Rouge (1988)',
 'Bent (1997)',
 'Big Bang Theory, The (1994)',
 'Bird of Prey (1996)',
 'Bonheur, Le (1965)',
 'Boys in Venice (1996)',
 "Brother's Kiss, A (1997)",
 'Careful (1992)',
 'Carmen Miranda: Bananas Is My Business (1994)',
 'Chairman of the Board (1998)',
 'Convent, The (Convento, O) (1995)',
 'Crows and Sparrows (1949)',
 'Crude Oasis, The (1995)',
 'Cyclo (1995)',
 'Daens (1992)',
 'Death in the Garden (Mort en ce jardin, La) (1956)',
 'Desert Winds (1995)',
 'Desperate Measures (1998)',
 'Entertaining Angels: The Dorothy Day Story (1996)',
 'Every Other Weekend (1990)',
 "Eye of Vichy, The (Oeil de Vichy, L') (1993)",
 'Fall (1997)',
 'Falling in Love Again (1980)',
 'Fear, The (1995)',
 'For Ever Mozart (1996)',
 'For Richer 